overlord/assertstate/bulk.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	db := cachedDB(s)
+
+	pool := asserts.NewPool(db, 512)
+
+	for instanceName, snapst := range snapStates {
+		sideInfo := snapst.CurrentSideInfo()
+		if sideInfo.SnapID == "" {
+			continue
+		}
+
+		declRef := &asserts.Ref{
+			Type:       asserts.SnapDeclarationType,
+			PrimaryKey: []string{release.Series, sideInfo.SnapID},
+		}
+		if err := pool.AddToUpdate(declRef, instanceName); err != nil {
+			return fmt.Errorf("cannot prepare snap-declaration refresh for snap %q: %v", instanceName, err)
+		}
+	}
+
+	modelAs := deviceCtx.Model()
+
+	// fetch store assertion if available
+	if modelAs.Store() != "" {
+		storeRef := asserts.Ref{
+			Type:       asserts.StoreType,
+			PrimaryKey: []string{modelAs.Store()},
+		}
+		if err := pool.AddToUpdate(&storeRef, storeGroup); err != nil {
+			if !asserts.IsNotFound(err) {
+				return fmt.Errorf("cannot prepare store assertion refresh: %v", err)
+			}
+			storeAt := &asserts.AtRevision{
+				Ref:      storeRef,
+				Revision: asserts.RevisionNotKnown,
+			}
+			err := pool.AddUnresolved(storeAt, storeGroup)
+			if err != nil {
+				return fmt.Errorf("cannot prepare store assertion fetching: %v", err)
+			}
+		}
+	}
+
+	err := resolvePool(s, pool, userID, deviceCtx)
+	if err != nil {
+		if rpe, ok := err.(*resolvePoolError); ok {
+			if e := rpe.errors[storeGroup]; asserts.IsNotFound(e) || e == asserts.ErrUnresolved {
+				// ignore
+				delete(rpe.errors, storeGroup)
+			}
+			if len(rpe.errors) == 0 {
+				return nil
+			}
+			rpe.message = "cannot refresh snap-declarations for snaps"
+		}
+	}
+	return err
+}
+
+// marker error to request falling back to the old implemention for assertion
+// refreshes
+var errBulkAssertionFallback = errors.New("bulk assertion request rejected or failed, fallback")
+
+type resolvePoolError struct {
+	message string
+	// errors maps groups to errors
+	errors map[string]error
+}
+
+func (rpe *resolvePoolError) Error() string {
+	message := rpe.message
+	if message == "" {
+		message = "cannot fetch and resolve assertions"
+	}
+	s := make([]string, 0, 1+len(rpe.errors))
+	s = append(s, fmt.Sprintf("%s:", message))
+	groups := make([]string, 0, len(rpe.errors))
+	for g := range rpe.errors {
+		groups = append(groups, g)
+	}
+	sort.Strings(groups)
+	for _, g := range groups {
+		s = append(s, fmt.Sprintf(" - %s: %v", g, rpe.errors[g]))
+	}
+	return strings.Join(s, "\n")
+}
+
+func resolvePool(s *state.State, pool *asserts.Pool, userID int, deviceCtx snapstate.DeviceContext) error {
+	user, err := userFromUserID(s, userID)
+	if err != nil {
+		return err
+	}
+	sto := snapstate.Store(s, deviceCtx)
+	db := cachedDB(s)
+	unsupported := handleUnsupported(db)
+
+	for {
+		// TODO: pass refresh options?
+		s.Unlock()
+		_, aresults, err := sto.SnapAction(context.TODO(), nil, nil, pool, user, nil)
+		s.Lock()
+		if err != nil {
+			if saErr, ok := err.(*store.SnapActionError); !ok || !saErr.NoResults || len(saErr.Other) != 0 {
```

Comment:
I found untangling the if condition slightly confusing. Maybe a type switch would make it easier?
```
switch stErr := err.(type) {
case *store.SnapActionError:
	if !stErr.NoResults || len(stErr.Other) != 0 {
		return errBulkAssertionFallback
	}
case *store.UnexpectedHTTPStatusError:
	if stErr.StatusCode >= 400 && stErr.StatusCode <= 500 {
		return errBulkAssertionFallback
	}
}
return err
```

Reply:
+1

Reply:
applied, changed though stError to stoErr, because "st" is used for/ sounds like "state"

Reply:
mmh, this change broke things (but not the unit tests), need to investigate

Reply:
I see, it's not quite equivalent, need to add a unit tests and see what to do

Reply:
I had to do this:  e81600247b945a3